### PR TITLE
Implement sync for Duck Address

### PR DESCRIPTION
**Pull request by cmonfortep**
Task/Issue URL: https://app.asana.com/0/488551667048375/1205449054184327/f 

### Description
Implements sync for Duck Address.

### Steps to test this PR

We will need 2 devices

_Sync between 2 devices_
- [ ] On first device, create a sync account
- [ ] Sign in on you Duck Address
- [ ] Go to your second device, login on the existing account
- [ ] Ensure you are now signed in on your Duck Address
- [ ] Visit a website where you can introduce an email (newsletters), and try to provide an alias, it should work
- [ ] on that device, sign out your duck address
- [ ] go to the first device, trigger sync (e.g: close and open the app)
- [ ] ensure you are now logged out from your Duck Address

_Merging account_
- [ ] Start with sync unauthenticated device, signed out from duck address
- [ ] Sign in you Duck Address
- [ ] Create a sync account
- [ ] ensure you are still signed in on your Duck Address account
- [ ] now copy your recovery code on the clipboard (settings -> sync -> show text code -> copy code)
- [ ] log out from your sync account
- [ ] sign out from you duck adress 
- [ ] Use the recovery code to login on your existing 
- [ ] Ensure you are now signed in on your Duck Address

### UI changes
| Before  | After |
| ------ | ----- |
!(Upload before screenshot)|(Upload after screenshot)|

### CODE REVIEW

**Comment by cmonfortep** [comment]
Current dependencies on/for this PR:
* develop
  * **PR #3522** <a href="https://app.graphite.dev/github/pr/duckduckgo/Android/3522" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/duckduckgo/Android/3522?utm_source=stack-comment).

File: settings/settings-impl/build.gradle [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    implementation project(path: ':anvil-annotations')
+
+    implementation AndroidX.appCompat
+    implementation Google.android.material
+    implementation JakeWharton.timber
+
+    implementation KotlinX.coroutines.core
+
+    implementation Square.retrofit2.converter.moshi
+    implementation "com.squareup.moshi:moshi-kotlin:_"
```
Comment (malmstein) [inline settings/settings-impl/build.gradle:43]:
```suggestion
    implementation Square.moshi"
```

File: settings/settings-impl/src/main/java/com/duckduckgo/settings/impl/SettingsSyncDataPersister.kt [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    ): SyncMergeResult {
+        if (changes.jsonString.isEmpty()) {
+            Timber.i("Sync-Settings: jsonString is empty")
+            return Success()
+        }
+
+        val response = runCatching {
+            updatesAdapter.fromJson(changes.jsonString)!!
+        }.getOrElse {
+            Timber.i("Sync-Settings: process() error parsing credentials ${it.message}")
```
Comment (malmstein) [inline settings/settings-impl/src/main/java/com/duckduckgo/settings/impl/SettingsSyncDataPersister.kt:74]: typo from copy paste?

File: settings/settings-impl/src/main/java/com/duckduckgo/settings/impl/SettingsSyncDataPersister.kt [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+                }
+            }
+        }
+
+        Timber.i("Sync-Settings: process() result=$result")
+        return result
+    }
+
+    private fun processEntries(
+        settings: CredentialsSyncEntries,
```
Comment (malmstein) [inline settings/settings-impl/src/main/java/com/duckduckgo/settings/impl/SettingsSyncDataPersister.kt:101]: Copy paste? Should be `SettingsSyncEntries`

**Review by malmstein — CHANGES_REQUESTED** [review]
Test cases work. But, as discussed in the task, I’ve found two issues. Also, we should discuss how you are using hard referenced lambdas to call one class from another instead of using DataObservers like in other syncable features.

**Review by malmstein — APPROVED** [review]
Discussed offline, all good!

File: app/build.gradle [hunk 4]
```diff
@@ -321,6 +324,7 @@ dependencies {
     implementation "com.squareup.okhttp3:okhttp-tls:_"
     implementation Square.retrofit2.retrofit
     implementation Square.retrofit2.converter.moshi
+    implementation "com.squareup.moshi:moshi-kotlin:_"
```
[the remark this answers is not in the document]
Comment (cmonfortep) [reply app/build.gradle:332]: Doesn't seem to be the right dependency. Tried `Square.moshi-kotlin` instead, but doesn't work. I'm not sure what's the short name for this one.

File: app/src/main/java/com/duckduckgo/app/email/AppEmailManager.kt [hunk 5]
```diff
@@ -45,11 +45,18 @@ import timber.log.Timber
 class AppEmailManager @Inject constructor(
     private val emailService: EmailService,
     private val emailDataStore: EmailDataStore,
+    private val emailSettings: EmailSync,
     private val dispatcherProvider: DispatcherProvider,
     private val appCoroutineScope: CoroutineScope,
     private val pixel: Pixel,
 ) : EmailManager, BrowserFeatureStateReporterPlugin {
 
+    init {
```
[the remark this answers is not in the document]
Comment (cmonfortep) [reply app/src/main/java/com/duckduckgo/app/email/AppEmailManager.kt]: Decided to move the discussion in a follow-up task.

Reasons why the current implementation:
- This class is emitting events in `signedInStateFlow` when another part of the code calls `storeCredentials` or `signOut`, but not based on the actual stored values.
- Sync applies changed directly at store level (directly in dao or sharedprefs)
- When that happens we need a way that emailmanager can signal a new event on that flow.
- Current approach, adds a listener to know when sync changes values.

File: app/src/main/java/com/duckduckgo/app/email/sync/EmailSync.kt [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+            val personalAccessToken = duckAddressSetting.personal_access_token
+            storeNewCredentials(duckAddress, personalAccessToken)
+            return true
+        } else {
+            storeNewCredentials("", "")
+            return true
+        }
+    }
+
+    override fun mergeRemote(value: String?): Boolean {
```
[the remark this answers is not in the document]
Comment (cmonfortep) [reply app/src/main/java/com/duckduckgo/app/email/sync/EmailSync.kt:66]: A syncable object should expose 2 different methods for storing data. First will be just to store value without applying any logic. Second will have business logic that decides how to merge local and remote data. The later happens when sync engine strategy is deduplicate.

File: settings/settings-impl/build.gradle (hunk 1 above)
Comment (cmonfortep) [reply settings/settings-impl/build.gradle:43]: ditto

File: settings/settings-impl/src/main/java/com/duckduckgo/settings/impl/SyncDateProvider.kt [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.duckduckgo.settings.impl
+
+import org.threeten.bp.*
+import org.threeten.bp.format.*
+
+class SyncDateProvider {
```
[the remark this answers is not in the document]
Comment (cmonfortep) [reply settings/settings-impl/src/main/java/com/duckduckgo/settings/impl/SyncDateProvider.kt:22]: While that's true, that belongs to sync implementation details. I'm going to defer this one to discuss how to include it in sync-module so features can rely on sync specific implementation detail.

I think we have similar things in other modules, and we can tackle them at the same time.

**Pull request merged**